include/ginkgo/core/solver/chebyshev.hpp

```diff
@@ -0,0 +1,245 @@
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_SOLVER_CHEBYSHEV_HPP_
+#define GKO_PUBLIC_CORE_SOLVER_CHEBYSHEV_HPP_
+
+
+#include <vector>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/identity.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/combined.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+#include <ginkgo/core/stop/iteration.hpp>
+
+
+namespace gko {
+namespace solver {
+
+
+/**
+ * Chebyshev iteration is an iterative method that can solving nonsymeetric
+ * problems. Chebyshev Iterations avoids the inner products for computation
+ * which may be the bottleneck for distributed system. Chebyshev Iteration is
+ * developed based on the Chebyshev polynomials of the first kind. Moreover,
+ * this method requires knowledge about the spectrum of the preconditioned
+ * matrix. This implementation follows the algorithm in "Templates for the
+ * Solution of Linear Systems: Building Blocks for Iterative Methods, 2nd
+ * Edition".
+ *
+ * ```
+ * solution = initial_guess
+ * while not converged:
+ *     residual = b - A solution
+ *     error = preconditioner(A) * residual
+ *     solution = solution + alpha_i * error + beta_i * (solution_i -
+ * solution_{i-1})
+ * ```
+ *
+ *
+ * @tparam ValueType  precision of matrix elements
+ *
+ * @ingroup solvers
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision>
+class Chebyshev
```

Comment:
nit: I would like to push making our classes final. I think I put down my reasons for this elsewhere.
```suggestion
class Chebyshev final
```